File: noaa-requests/noaa-py/noaapy/wl_download.py
# noaa-requests/noaa-py/noaapy/wl_download.py
import time
from io import StringIO
from typing import Any, Dict, List

import numpy as np
import pandas as pd
import requests


def _download_csv_as_dataframe(
    url: str, options: Dict[str, Any] | None = None
) -> pd.DataFrame:
    """
    Helper to download a CSV from NOAA CO-OPS and return as a pandas DataFrame.
    """
    opts = options.copy() if options else {}
    timeout = opts.pop("timeout", 30)  # pull out timeout if provided

    resp = requests.get(url, timeout=timeout, **opts)
    resp.raise_for_status()
    return pd.read_csv(StringIO(resp.text))


def wl_download(
    ii: int,
    Sdata: List[dict],
    datum: str,
    station: str,
    timezone: str,
    units: str,
    fmt: str,  # 'format' is a Python builtin, so use fmt instead
    stDates,
    endDates,
    gen_url: str,
    options: Dict[str, Any] | None,
    flag1: str,
) -> List[dict]:
    """
    Python translation of MATLAB WL_download.m

    Parameters
    ----------
    ii : int
        Index into Sdata (0-based in Python: Sdata[ii]).
    Sdata : list of dict
        Sdata[ii]['WL'] will be set to the resulting DataFrame.
    datum, station, timezone, units, fmt, gen_url, flag1 :
        Same semantics as in the MATLAB function.
    stDates, endDates :
        For non-monthly:
            - list of list of strings, e.g. stDates[jj][kk] == 'yyyymmdd HH:MM'
        For monthly:
            - list (or 1-D array) of strings 'yyyymmdd HH:MM' or 'yyyymmdd'.
    options : dict or None
        Extra options for requests.get (e.g. {'timeout': 30}).

    Returns
    -------
    Sdata : list of dict
        Same object as input, with Sdata[ii]['WL'] set to a DataFrame
        with columns ['DateTime', 'WaterLevel'].
    """
    data = pd.DataFrame(columns=["DateTime", "WaterLevel"])

    # Determine length of outer loop
    if flag1 != "m":
        Lend = len(stDates)
    else:
        # MATLAB used length(stDates(:,1)); here we assume stDates is 1-D
        Lend = len(stDates)

    for jj in range(Lend):
        # GAP FILLER (placeholder)
        data = gap_filler(data, stDates, endDates, flag1, jj)

        # ------------------------------------------------------------------
        # NON-MONTHLY PRODUCTS (6-min, hourly, high/low)
        # ------------------------------------------------------------------
        if flag1 != "m":
            # stDates[jj] and endDates[jj] should be lists of 'yyyymmdd HH:MM' strings
            segments = stDates[jj]
            end_segments = endDates[jj]

            for kk in range(len(segments)):
                begin_str = segments[kk]  # 'yyyymmdd HH:MM'
                end_str = end_segments[kk]

                if flag1 in ("6", "6p"):
                    # 6-minute WL measurements
                    wl_api = (
                        "/api/prod/datagetter?product=water_level"
                        "&application=NOS.COOPS.TAC.WL"
                        f"&begin_date={begin_str}"
                        f"&end_date={end_str}"
                        f"&datum={datum}"
                        f"&station={station}"
                        f"&time_zone={timezone}"
                        f"&units={units}"
                        f"&format={fmt}"
                    )

                    url = gen_url + wl_api

                    # Retry loop (as in MATLAB try/catch + while)
                    wltable = None
                    while wltable is None:
                        try:
                            wltable = _download_csv_as_dataframe(url, options)
                        except Exception:
                            time.sleep(3)

                elif flag1 == "1":
                    # 1-hour WL measurements
                    wl_api = (
                        "/api/prod/datagetter?product=hourly_height"
                        "&application=NOS.COOPS.TAC.WL"
                        f"&begin_date={begin_str}"
                        f"&end_date={end_str}"
                        f"&datum={datum}"
                        f"&station={station}"
                        f"&time_zone={timezone}"
                        f"&units={units}"
                        f"&format={fmt}"
                    )
                    url = gen_url + wl_api
                    wltable = _download_csv_as_dataframe(url, options)

                elif flag1 == "hilo":
                    # High/low WL measurements
                    wl_api = (
                        "/api/prod/datagetter?product=high_low"
                        "&application=NOS.COOPS.TAC.WL"
                        f"&begin_date={begin_str}"
                        f"&end_date={end_str}"
                        f"&datum={datum}"
                        f"&station={station}"
                        f"&time_zone={timezone}"
                        f"&units={units}"
                        f"&format={fmt}"
                    )
                    url = gen_url + wl_api
                    wltable = _download_csv_as_dataframe(url, options)

                else:
                    raise ValueError(f"Unsupported flag1 for non-monthly data: {flag1}")

                # ------------------------------------------------------------------
                # FAILSAFE: if empty or "too many" columns, synthesize NaN series
                # ------------------------------------------------------------------
                if wltable.empty or wltable.shape[1] > 5:
                    # MATLAB:
                    # uBound = datenum(stDates{jj}{kk},'yyyymmdd HH:MM');
                    # lBound = datenum(endDates{jj}{kk},'yyyymmdd HH:MM');
                    u_bound = pd.to_datetime(begin_str, format="%Y%m%d %H:%M")
                    l_bound = pd.to_datetime(end_str, format="%Y%m%d %H:%M")

                    # timestep
                    if "6" in flag1:
                        freq = "6min"
                    else:
                        freq = "1H"

                    # datetime vector from uBound to lBound, then drop the first
                    datetimes = pd.date_range(start=u_bound, end=l_bound, freq=freq)[1:]

                    # Create NaN vector
                    water_level = np.full(shape=len(datetimes), fill_value=np.nan)

                    # Synthetic table
                    wltable = pd.DataFrame(
                        {"DateTime": datetimes, "WaterLevel": water_level}
                    )
                else:
                    # Take first two columns (datetime, water level) like wltable(:,1:2)
                    # We rename them to ['DateTime','WaterLevel'] for consistency.
                    # NOAA CSV typically has 'Date Time' and 'Water Level' etc.
                    subset = wltable.iloc[:, :2].copy()
                    subset.columns = ["DateTime", "WaterLevel"]

                    # Convert DateTime to pandas datetime if it's not already
                    subset["DateTime"] = pd.to_datetime(subset["DateTime"])
                    wltable = subset

                print(
                    f"Station: {station} WL Measurements: {jj + 1}/{Lend} "
                    f"Segmentation: {kk + 1}/{len(segments)}"
                )

                # Concatenate chunk
                data = pd.concat([data, wltable], ignore_index=True)

        # ------------------------------------------------------------------
        # MONTHLY PRODUCT
        # ------------------------------------------------------------------
        else:
            # Monthly WL measurements
            begin_str = stDates[jj]  # assume 'yyyymmdd' or 'yyyymmdd HH:MM'
            end_str = endDates[jj]

            wl_api = (
                "/api/prod/datagetter?product=monthly_mean"
                "&application=NOS.COOPS.TAC.WL"
                f"&begin_date={begin_str}"
                f"&end_date={end_str}"
                f"&datum={datum}"
                f"&station={station}"
                f"&time_zone={timezone}"
                f"&units={units}"
                f"&format={fmt}"
            )

            url = gen_url + wl_api
            wltable = _download_csv_as_dataframe(url, options)

            # NOAA monthly CSV typically has a 'Date' column like 'YYYY-MM', and 'MSL'.
            # We map this into the same DateTime/WaterLevel structure and fill gaps.
            if "Date" not in wltable.columns:
                raise ValueError(
                    "Expected a 'Date' column in monthly_mean response; "
                    f"got columns: {list(wltable.columns)}"
                )
            if "MSL" not in wltable.columns:
                raise ValueError(
                    "Expected an 'MSL' column in monthly_mean response; "
                    f"got columns: {list(wltable.columns)}"
                )

            wltable["DateTime"] = pd.to_datetime(wltable["Date"])
            wltable["WaterLevel"] = wltable["MSL"]

            # Set monthly frequency and fill gaps with NaNs (like the MATLAB loop)
            monthly = (
                wltable[["DateTime", "WaterLevel"]]
                .set_index("DateTime")
                .asfreq("MS")  # Month-start frequency
            )

            monthly = monthly.reset_index()  # back to columns
            monthly.columns = ["DateTime", "WaterLevel"]

            # Concatenate chunk
            data = pd.concat([data, monthly], ignore_index=True)

    # ----------------------------------------------------------------------
    # REMOVE DUPLICATE POINTS (MEASUREMENTS)
    # ----------------------------------------------------------------------
    # Sort by DateTime, then drop duplicates keeping the last (as MATLAB unique(...,'last'))
    data = data.sort_values("DateTime").drop_duplicates(subset="DateTime", keep="last")

    # Attach to Sdata and return
    if ii >= len(Sdata):
        # extend Sdata if needed
        Sdata.extend({} for _ in range(ii - len(Sdata) + 1))

    Sdata[ii]["WL"] = data.reset_index(drop=True)
    return Sdata

File: conversion/noaa-requests/matlab/NOAA_Water_Level_Downloader/WL_Downloader_Functions/WL_download.m
%{
This function is responsible for downloading WL mwasurements from NOAA CO-OPS
server.
%}
%% BUILD URL & DOWNLOAD MEASURED DATA
function [Sdata] = WL_download(ii,Sdata,datum,station,timezone,units,format,stDates,endDates,gen_url,options,flag1)
data=[];
% For Loop length (Monthly product code does not work with cell arrays)
 if strcmp(flag1,'m')==0
     Lend = length(stDates);
 else 
     Lend = length(stDates(:,1));
 end

for jj = 1:Lend
    %% NAN GAP FILLER
    [data] = gap_filler(data,stDates,endDates,flag1,jj);
    
    %% DATA DOWNLOAD
    if strcmp(flag1,'m')==0
        %% hourly and 6 minutes data
        for kk = 1:length(stDates{jj})
            if strcmp(flag1,'6')==1 || strcmp(flag1,'6p')==1 
                % 6 Minute WL Measurements API
%                 wl6mn_api = ['/api/datagetter?product=water_level&application=NOS.COOPS.TAC.WL&begin_date=',stDates{jj}{kk},...
%                     '&end_date=',endDates{jj}{kk},'&datum=',datum,'&station=',station,'&time_zone=',timezone,'&units=',units,'&format=',format,''];
                wl6mn_api = ['/api/prod/datagetter?product=water_level&application=NOS.COOPS.TAC.WL&begin_date=',stDates{jj}{kk},...
                    '&end_date=',endDates{jj}{kk},'&datum=',datum,'&station=',station,'&time_zone=',timezone,'&units=',units,'&format=',format,''];

                % Cat URL
                URL = [gen_url wl6mn_api];
                % Donwload data
                try
                    wltable = webread(URL,options);                        
                catch
                    wexit = 0;
                    while wexit==0
                        pause(3);
                        try
                            dmmy=0;
                            wltable = webread(URL,options);
                        catch
                            dmmy = 1;
                        end
                        if dmmy==0
                            wexit = 1;
                        end
                    end
                end
            elseif strcmp(flag1,'1')==1
                % 1 Hour WL Measurements API
                wl1hr_api = ['/api/prod/datagetter?product=hourly_height&application=NOS.COOPS.TAC.WL&begin_date=',stDates{jj}{kk},...
                    '&end_date=',endDates{jj}{kk},'&datum=',datum,'&station=',station,'&time_zone=',timezone,'&units=',units,'&format=',format,''];
                % Cat URL
                URL = [gen_url wl1hr_api];
                % Download data
                wltable = webread(URL,options);
            elseif strcmp(flag1,'hilo')==1
                              % 1 Hour WL Measurements API
                wl1hr_api = ['/api/prod/datagetter?product=high_low&application=NOS.COOPS.TAC.WL&begin_date=',stDates{jj}{kk},...
                    '&end_date=',endDates{jj}{kk},'&datum=',datum,'&station=',station,'&time_zone=',timezone,'&units=',units,'&format=',format,''];
                % Cat URL
                URL = [gen_url wl1hr_api];
                % Download data
                wltable = webread(URL,options);  
            end
            % Temporary failsafe - got empty table when there is data in
            % record 
             if isempty(wltable)==1 || width(wltable)>5
                 uBound = datenum(stDates{jj}{kk},'yyyymmdd HH:MM');
                 % Get Initial Time Stamp of Actual Segmentation
                 lBound = datenum(endDates{jj}{kk},'yyyymmdd HH:MM');
                 % Define timestep
                 if contains(flag1,'6')==1
                     dt = datenum(0,0,0,0,6,0);
                 else
                     dt = datenum(0,0,0,1,0,0);
                 end
                 % Create datetime vector
                 DateTime = datetime(datestr([uBound:dt:lBound]','yyyy-mm-dd HH:MM'),...
                     'InputFormat','yyyy-MM-dd HH:mm','Format','yyyy-MM-dd HH:mm');
                 % Eliminate uBound value
                 DateTime = DateTime(2:end);
                 % Create NaN vector
                 WaterLevel = NaN(size(DateTime));
                 % Create fill table
                 wltable = table(DateTime,WaterLevel);
             end
             ['Station: ',station,' WL Measurements: ',num2str(jj),'/',num2str(Lend),' Segmentation: ',num2str(kk),'/',num2str(length(stDates{jj})),'']
            % Cat downloaded data chunk
            data = [data;wltable(:,1:2)];
        end
    else
        %% Monthly data only
        % Monthly WL Measurements API
        wlm_api = ['/api/prod/datagetter?product=monthly_mean&application=NOS.COOPS.TAC.WL&begin_date=',stDates(jj,:),...
            '&end_date=',endDates(jj,:),'&datum=',datum,'&station=',station,'&time_zone=',timezone,'&units=',units,'&format=',format,''];
        % Cat URL
        URL = [gen_url wlm_api];
        % Download data
        wltable = webread(URL,options);
        %% MONTHLY DATE VECTOR FORMATER
        % Counters and triggers
        dd=0;ctr=1;ctr2=1;
        % Clear vars
        clearvars('DateTime','WaterLevel');
        % Conversion while loop
        if height(wltable)==1
                    % Date assignment and formating
                DateTime(1,1) = datetime(wltable.Year(1),wltable.Month(1),1,'Format','yyyy-MM-dd HH:mm');
                % Water level assignment
                WaterLevel(1,1) = wltable.MSL(1);
        else
        while dd==0
            % Define if triggers
            chk = wltable.Month(ctr2);
            chk2 = wltable.Month(ctr2+1);
            % If data is continous assign date to new vector
            if chk2-chk==1 || chk2-chk==-11
                % Date assignment and formating
                DateTime(ctr,1) = datetime(wltable.Year(ctr2),wltable.Month(ctr2),1,'Format','yyyy-MM-dd HH:mm');
                % Water level assignment
                WaterLevel(ctr,1) = wltable.MSL(ctr2);
                % Counter increments
                ctr2=ctr2+1;
                ctr=ctr+1;
            else
                % If data is not continous add missing dates
                % Assing last date before finding gap
                DateTime(ctr,1) = datetime(wltable.Year(ctr2),wltable.Month(ctr2),1,'Format','yyyy-MM-dd HH:mm');
                WaterLevel(ctr,1) = wltable.MSL(ctr2);
                ctr=ctr+1;
                % Compute how many months need to be filled in
                dc = split(between(datetime(wltable.Year(ctr2),wltable.Month(ctr2),1),...
                    datetime(wltable.Year(ctr2+1),wltable.Month(ctr2+1),1)),'months');
                % Loop corrector
                for ll = 1:dc-1
                    DateTime(ctr,1) = datetime(wltable.Year(ctr2),wltable.Month(ctr2)+ll,1,'Format','yyyy-MM-dd HH:mm');
                    WaterLevel(ctr,1) = NaN(1,1);
                    ctr=ctr+1;
                end
                % Increment counter
                ctr2=ctr2+1;
            end
            % While exit condition
            if ctr2==height(wltable)
                % Define exit trigger
                dd=1;
                % Assign lasta data point
                DateTime(ctr,1) = datetime(wltable.Year(ctr2),wltable.Month(ctr2),1,'Format','yyyy-MM-dd HH:mm');
                WaterLevel(ctr,1) = wltable.MSL(ctr2);
            end
        end
        end
        % Cat downloaded data chunk
        data = [data;table(DateTime,WaterLevel)];
    end
end

%% REMOVE DUPLICATE POINTS (MEASUREMENTS)
Sdata(ii).WL = data;
[~,uIndx] = unique(Sdata(ii).WL(:,1),'last');
Sdata(ii).WL = Sdata(ii).WL(uIndx,:);
end